### PR TITLE
Add SHA512

### PR DESCRIPTION
**Pull request by varunthakore**
`bellpepper-sha512` crate implements circuit for SHA-512 hash function and circuit representation of u64.

### CODE REVIEW

**Comment by avras** [comment]
I have reviewed this code before, when Varun submitted the SHA512 gadget to the ZKP MOOC hackathon. 

It's ready to be merged.

**Review by avras — DISMISSED** [review]

File: crates/bellpepper-sha512/src/sha512.rs [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+    let h7 = UInt64::addmany(
+        cs.namespace(|| "new h7"),
+        &[current_hash_value[7].clone(), h],
+    )?;
+
+    Ok(vec![h0, h1, h2, h3, h4, h5, h6, h7])
+}
+
+#[cfg(test)]
+mod test {
```
Comment (huitseeker) [inline crates/bellpepper-sha512/src/sha512.rs:353]: The tests which run quickly would benefit from running on more than 1 input, consider adding, like in the u64 tests, a
```rust
for _ in 1..100 {
   <test content>
}
```
inner loop. Indicative runtimes in opt mode:
```
        PASS [   0.156s] bellpepper-sha512 sha512::test::test_hash
        PASS [   0.164s] bellpepper-sha512 sha512::test::test_full_block
        PASS [   0.236s] bellpepper-sha512 uint64::test::test_uint64_sha512_maj
        PASS [   0.241s] bellpepper-sha512 uint64::test::test_uint64_sha512_ch
        PASS [   0.297s] bellpepper-sha512 sha512::test::test_full_hash
        PASS [  29.519s] bellpepper-sha512 sha512::test::test_against_vectors
```

See https://github.com/lurk-lab/bellpepper-gadgets/pull/8 for how those numbers are relevant (setting an optimization level of 3 for CI)

**Comment by avras** [comment]
Thanks @huitseeker for the comments. Sorry about the missing the right SHA-512 reference in the README.

@varunthakore Can you also change the directory name from `bellperson-sha512` to `sha512`? Like the changes in https://github.com/lurk-lab/bellpepper-gadgets/pull/5. Thanks.

**Comment by varunthakore** [comment]
Changed the directory name to `sha512` and made changes as suggested by @huitseeker.

File: crates/sha512/src/sha512.rs [hunk 2]
```diff
@@ -0,0 +1,512 @@
+//! Circuits for the [SHA-512] hash function and its internal compression
+//! function.
+//!
+//! [SHA-512]: https://tools.ietf.org/html/rfc6234
+
+// #![allow(clippy::many_single_char_names)]
```
Comment (huitseeker) [inline crates/sha512/src/sha512.rs:6]: Could you remove this comment?

**Review by huitseeker — COMMENTED** [review]
Thanks a bunch! Just one cosmetic change, and we can get this one in!

**Review by huitseeker — APPROVED** [review]
🙏 Thanks a lot!!

**Pull request merged**